### DISCUSSION TITLE
Naming things

### DISCUSSION THREAD

**Discussion by wkitlasten** [gmdsi/GMDSI_notebooks#108] [General]
Hey, I need to get a better handle on the par and obs (other?) naming conventions. The few "special characters" I know I can make better use of (as mentioned in the notebooks) are:

: signifies equals or "value to follow"
_ break names into components

So
`inst:0_ptype:pp_pstyle:a_i:9_j:75_zone:1`
which translates to a parameter with the following characteristics:
```
instance = 0
ptype = pp
pstyle = a
i = 9
j=75
zone = 1
```

Obviously I need to avoid '_' and ':' in par and obs names. But it also seems wise to avoid `.`, `-`, and obviously `,` which can wreak havoc on csv formated instruction files (I'm a heuristic learner). For example, while obvious in `gage-1`, the `-` might get confused with the negative somewhere down the line?

[Note: We have used a naming convention for our observations. Each observation name starts with the site name (e.g. "gage-1"), followed by " : " and then the simulation time in days (e.g. "1.0"). So, "gage_1:1.0" refers to the observation at "gage-1" after "1.0" days (in this case, at the end of the first steady state stress period).](https://github.com/gmdsi/GMDSI_notebooks/blob/a7805ad1916fbebfdd4651a66f0f1b00ecdc396e/tutorials/part1_03_calibrate_k/freyberg_k.ipynb)

Also, using `.` in par and obs names may cause silly people like me issues when setting up pest files based on file names (e.g. `model.ghb.cond.csv` and `model.ghb.elev.csv` files to make `ghb.cond` and `ghb.elev` based pars).

Any other special flags that should be avoided and/or taken advantage of while I formulate my universal, all encompassing, and ultimate-final naming conventions? 
```
,
.
/
\
[
]
etc?
```

**Comment by laat0003** [comment]
Yeah. I asked a question about something similar to this in the first session because I was concerned that some of the inbuilt plotting or data-processing utilities expected certain naming, conventions. I use the 'tag' method that is demonstrated in the notebooks to slice dataframes when I run my post processing and have experienced issues (unrelated to Pyemu itself) with dashes and periods in obs names. Creative use of string methods like split(), strip(), lstrip() and rstrip() usually save the day but also means I have to edit scripts constantly. I'm all for a "10 commandments" style approach to naming conventions for par, obs and their groups, so keen to hear from the devs myself on this one.

**Comment by jtwhite79** [comment]
Well my scheme this week (subject to change) is to cast str to lower case, strip leading and trailing whitespace, and then replace all underscores, internal whitespace(s), and other rogue characters in row and column labels with hyphens; I also to store datetimes as "d-m-y", basically trying to stay away from colon and underscore since pyemu uses those to parse metadata from names.  I'm also a tilde guy for the tpl and ins file markers, so I steer clear of that character.  

@wkitlasten so far Ive not run into the hyphen/negative sign confusion, but when it happens, Im sure it will take several days figure out what happened and then a new, equally haphazard naming scheme will be adopted! haha.

For par groups, I usually just go with the groups that `PstFrom` comes up with.  But for obs groups, I usually do modify those after `PstFrom`.  This could be because I want to use inequality-type obs or Im using the internal IES weight adjustment process that relies on obs group names.  

Im keen to hear other people's schemes - Im sure they are less convoluted...or if anyone has a suggestion on how to improve this metadata-in-names idea...

**Comment by laat0003** [comment]
Cheers for that @jtwhite79. It's nice to know that you change schemes too. I find that every project throws in a naming convention that causes issues for my scripts with some names. I've just been using whatever the defaults are with Pyemu for par, obs, groups and then band-aid whatever fails as needed but have not really put the mental effort into coming up with a scheme of my own thus far. I guess I was being lazy and pinning my hopes on smarter people than me. :)

**Comment by wkitlasten** [comment]
"The hardest part about coding is naming things." Mike Taves

The biggest (current) nuisance for me is the use of `_` to break down the components of the long names (e.g., `i:1_j:2`) since the columns in my df, and hence my `use_cols` arg tend to have `_`. All my "useful" metadata turns into meaningless jibberish... 

What about reserving `_` as a substitute for spaces in names and using `;` to break the components of the long names? Unless it could be interpreted as a "statement separator" somewhere in the code? Or perhaps reserving `_` as a substitute for spaces, `:` to break components, and the more traditional `=` for "value to follow" as in:

`dum_name_1=1;dum_name_2=2`

That'd be my vote... at least today.
